### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by camfitz**
Refactor, fix all the things.

Some limitations of the FileSystem class are noted and will be revised after basic Git functions are operational.

### CODE REVIEW

File: gitnu/app/gitnu.dart [hunk 1]
```diff
@@ -1 +1 @@
...
    if(emWidth > window.innerWidth ~/ 2) {
      colCount = 1;
    } else if(emWidth > window.innerWidth ~/ 3) {
      colCount = 2;
    }
    
    sb.write('<div class="ls-files" style="-webkit-column-count: $colCount;">');
    return sb;
  }
}
```
Comment (soulplant) [inline gitnu/app/gitnu.dart:1]: Hm, github seems to think this file is all on one line.

> ```
> file = new GitnuFileSystem("#file_path");
> ```
> 
> file -> filesystem

nits: if(...) -> if (...)

In formatColumns() could you explain the magic number 14 through comment / putting it in a constant.

File: gitnu/app/gitnufilesystem.dart [hunk 2]
```diff
@@ -0,0 +1,404 @@
+library GitnuFileSystem;
+
+import 'dart:html';
+import 'dart:async';
+import 'package:chrome_gen/chrome_app.dart' as chrome;
+import 'statictoolkit.dart';
+
+class GitnuFileSystem {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:8]: Add a class comment.

File: gitnu/app/gitnufilesystem.dart [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+  DirectoryEntry root;
+  DirectoryEntry cwd;
+  
+  GitnuFileSystem(String displayFilePath) {
+    // Check if we set a rootFolder in a previous use of the app.
+    chrome.storage.local.get('rootFolder').then((items) {
+      chrome.fileSystem.isRestorable(items["rootFolder"]).then((value) {
+        if(value == true) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:16]: nit: space between if and parens

Also, can this just be written "if (value) {" ?

File: gitnu/app/gitnufilesystem.dart [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+import 'package:chrome_gen/chrome_app.dart' as chrome;
+import 'statictoolkit.dart';
+
+class GitnuFileSystem {
+  DirectoryEntry root;
+  DirectoryEntry cwd;
+  
+  GitnuFileSystem(String displayFilePath) {
+    // Check if we set a rootFolder in a previous use of the app.
+    chrome.storage.local.get('rootFolder').then((items) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:14]: nit: Make 'rootFolder' a constant, eg: kRootFolder, and use it instead of the literal throughout this file.

File: gitnu/app/gitnufilesystem.dart [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+        } else {
+          window.console.debug("No root folder to restore."); 
+        }
+      });
+    });
+  }
+  
+  /*
+   * Opens a folder window allowing you to choose a folder.
+   * Handler is returned, and the entry is retained in chrome.storage -> input storage name.
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:38]: I don't think anything is returned from this function as it's asynchronous.

Also, could you rephrase the second part of the sentence as "The entry is retained in chrome.storage.local, keyed by |storageName|."

File: gitnu/app/gitnufilesystem.dart [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+   */
+  void openHandler(String storageName) {
+    chrome.ChooseEntryOptions options = new chrome.ChooseEntryOptions(
+        type: chrome.ChooseEntryType.OPEN_DIRECTORY);
+    chrome.fileSystem.chooseEntry(options).then((chrome.ChooseEntryResult res) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:43]: nit: Favour full words rather than abbreviations, ie: res -> result.

File: gitnu/app/gitnufilesystem.dart [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+      DirectoryEntry theEntry = res.entry;
+      
+      // use local storage to retain access to this file
+      chrome.storage.local.set({storageName: chrome.fileSystem.retainEntry(theEntry)}).then((storageArea) {
+        window.console.debug("Retained chosen folder- " + theEntry.fullPath + " as " + storageName);    
+      });
+      
+      if(storageName == "rootFolder") {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:51]: nit: space between if and parens

Also, is this ever called with storageName != "rootFolder"? If not you should remove this if statement, and if so then you should factor the code so that this function is generic (ie: returns a future with the entry in it) and have callers decide what to do with the result themselves, rather than passing in a control argument like this.

File: gitnu/app/gitnufilesystem.dart [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+   * Handler is returned, and the entry is retained in chrome.storage -> input storage name.
+   */
+  void openHandler(String storageName) {
+    chrome.ChooseEntryOptions options = new chrome.ChooseEntryOptions(
+        type: chrome.ChooseEntryType.OPEN_DIRECTORY);
+    chrome.fileSystem.chooseEntry(options).then((chrome.ChooseEntryResult res) {
+      DirectoryEntry theEntry = res.entry;
+      
+      // use local storage to retain access to this file
+      chrome.storage.local.set({storageName: chrome.fileSystem.retainEntry(theEntry)}).then((storageArea) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:47]: chrome.storage.local.set() returns nothing, I think, so "storageArea" is unnecessary.

File: gitnu/app/gitnufilesystem.dart [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+        }
+      });
+    });
+  }
+  
+  /*
+   * Opens a folder window allowing you to choose a folder.
+   * Handler is returned, and the entry is retained in chrome.storage -> input storage name.
+   */
+  void openHandler(String storageName) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:40]: openHandler() isn't very descriptive, maybe something like promptUserForNewRoot()

File: gitnu/app/gitnufilesystem.dart [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+          });
+        } else {
+          window.console.debug("No root folder to restore."); 
+        }
+      });
+    });
+  }
+  
+  /*
+   * Opens a folder window allowing you to choose a folder.
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:37]: "Shows a folder picker prompt, allowing the user to choose a new root folder."

File: gitnu/app/gitnufilesystem.dart [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+   * Handler is returned, and the entry is retained in chrome.storage -> input storage name.
+   */
+  void openHandler(String storageName) {
+    chrome.ChooseEntryOptions options = new chrome.ChooseEntryOptions(
+        type: chrome.ChooseEntryType.OPEN_DIRECTORY);
+    chrome.fileSystem.chooseEntry(options).then((chrome.ChooseEntryResult res) {
+      DirectoryEntry theEntry = res.entry;
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:44]: theEntry -> entry

File: gitnu/app/gitnufilesystem.dart [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+      
+      // use local storage to retain access to this file
+      chrome.storage.local.set({storageName: chrome.fileSystem.retainEntry(theEntry)}).then((storageArea) {
+        window.console.debug("Retained chosen folder- " + theEntry.fullPath + " as " + storageName);    
+      });
+      
+      if(storageName == "rootFolder") {
+        this.root = theEntry;
+        
+        // Change the current working directory as well.
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:54]: This comment is unnecessary (and then remove the newline so the two assignments are next to each other).

File: gitnu/app/gitnufilesystem.dart [hunk 13]
```diff
@@ -0,0 +1,404 @@
+library GitnuFileSystem;
+
+import 'dart:html';
+import 'dart:async';
+import 'package:chrome_gen/chrome_app.dart' as chrome;
+import 'statictoolkit.dart';
+
+class GitnuFileSystem {
+  DirectoryEntry root;
+  DirectoryEntry cwd;
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:10]: Add comments for these variables.

File: gitnu/app/gitnufilesystem.dart [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+      case FileError.INVALID_MODIFICATION_ERR:
+        display('${StaticToolkit.htmlEscape(cmd)}: ${StaticToolkit.htmlEscape(dest)}: File already exists<br>');
+        break;
+      default:
+        errorHandler(error);
+        break;
+    }
+  }
+  
+  void read(String cmd, String path, Function display, var callback) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:81]: Add a proper type for |callback| and add a comment for this function.

File: gitnu/app/gitnufilesystem.dart [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+    cwd.getFile(path).then((FileEntry fileEntry) {
+      fileEntry.file().then((file) {
+        var reader = new FileReader();
+        reader.onLoadEnd.listen((ProgressEvent event) => callback(reader.result));
+        reader.readAsText(file);
+      }, onError: errorHandler);
+    }, onError: (error) {
+      if (error.code == FileError.INVALID_STATE_ERR) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:89]: Can you call invalidOpForEntryType() here?

File: gitnu/app/gitnufilesystem.dart [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+        msg = 'INVALID_MODIFICATION_ERR';
+        break;
+      case FileError.INVALID_STATE_ERR:
+        msg = 'INVALID_STATE_ERR';
+        break;
+      case FileError.TYPE_MISMATCH_ERR:
+        msg = 'TYPE_MISMATCH_ERR';
+        break;
+      default:
+        msg = 'FileError = ${error.code}: Error not handled';
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:121]: Error not handled -> Unknown error.

File: gitnu/app/gitnufilesystem.dart [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+        break;
+    };
+    // Need to work out how best to output errors!
+    window.console.debug('Error: ${StaticToolkit.htmlEscape(msg)}');
+  }
+  
+  void catCommand(String cmd, List<String> args, Function display) {
+    if (args.length >= 1) {
+      var fileName = args[0];
+      read(cmd, fileName, display, (result) => display('<pre>${StaticToolkit.htmlEscape(result)}</pre>'));
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:131]: Rather than passing in display functions to everything, it might be better to define a TerminalOutput interface or similar that looks like...

```
interface TerminalOutput {
  printLine(String line);
  printTable(List<String> items);
  // ...
}
```

Modulo syntax, I don't know dart :-P

The idea being that a TerminalOutput would be an instance variable of the class, which can be referenced by any method. This would also provide an interface that can be implemented in tests that will let you query the logical structure of the output, rather than needing to parse HTML, ie:

```
class FakeTerminalOutput extends TerminalOutput {
  printLine(String line) {
    this.lines.add(line);
  }
  printTable(List<String> items) {
    this.lines.add(items.join(" "));
  }
}

class HtmlTerminalOutput extends TerminalOutput {
  printLine(String line) {
    // html-escape and <pre>-ify |line| here, and append it to the DOM.
  }
}
```

File: gitnu/app/gitnufilesystem.dart [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+   */
+  void changeDirectory(List<String> args, Function display) {
+    var dest = args.join(' ').trim();
+    if(dest.isEmpty) {
+      dest = '/';
+    }
+    
+    cwd.getDirectory(dest)
+      .then((DirectoryEntry dirEntry) {
+        cwd = dirEntry;
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:148]: This should probably be encapsulated in a method setCurrentDirectory() (you set the cwd and update the display above in the constructor, so both places should just call setCurrentDirectory()).

This method could be called cdCommand() to be consistent with catCommand() (and listDirectory() could be lsCommand() below).

File: gitnu/app/gitnufilesystem.dart [hunk 19]
```diff
@@ -0,0 +1,404 @@
...
+      display('usage: ${StaticToolkit.htmlEscape(cmd)} filename');
+    }
+  }
+  
+  /**
+   * cd calls this.
+   */
+  void changeDirectory(List<String> args, Function display) {
+    var dest = args.join(' ').trim();
+    if(dest.isEmpty) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:142]: nit: space between if and parens

File: gitnu/app/gitnufilesystem.dart [hunk 20]
```diff
@@ -0,0 +1,404 @@
...
+    DirectoryReader reader = cwd.createReader();
+    
+    void readEntries() {
+      reader.readEntries().then((List<Entry> results) {
+        if (results.length == 0) {
+          entries.sort((a, b) => a.name.compareTo(b.name));
+          display(entries);
+        } else {
+          entries.addAll(results);
+          window.console.debug("readEntries call");
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:171]: remove debugging

File: gitnu/app/gitnufilesystem.dart [hunk 21]
```diff
@@ -0,0 +1,404 @@
...
+          readEntries();
+        }
+      }, onError: errorHandler);
+    };
+    
+    readEntries();
+  }
+  
+  void createDir(DirectoryEntry rootDirEntry, List<String> folders, Function display, [String createFromDir="", String cmd=""]) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:180]: nit: wrap at 80 chars

And add a comment for this function.

Also, createDir() -> createDirectory().

And I don't think it's worth passing down cmd just for error reporting.

File: gitnu/app/gitnufilesystem.dart [hunk 22]
```diff
@@ -0,0 +1,404 @@
...
+    if (folders.length == 0) {
+      return;
+    }
+
+    if (createFromDir.isEmpty) {
+      rootDirEntry.createDirectory(folders[0])
+      .then((dirEntry) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:187]: nit: elsewhere you put .then on the previous line, which I think is easier to read (here and elsewhere in this function).

File: gitnu/app/gitnufilesystem.dart (hunk 8 above)
Comment (camfitz) [inline gitnu/app/gitnufilesystem.dart:47]: This one is strange. It won't let me not have a value in brackets here... I've indicated it is unused with an underscore.

File: gitnu/app/gitnufilesystem.dart (hunk 15 above)
Comment (camfitz) [inline gitnu/app/gitnufilesystem.dart:89]: Only for the second error type - the first error type results in the opposite comment. Changed.

File: gitnu/app/gitnufilesystem.dart [hunk 23]
```diff
@@ -0,0 +1,404 @@
...
+        if (folders.length != 0) {
+          folders.removeAt(0);
+          createDir(dirEntry, folders, display);
+        }
+      }, onError: errorHandler);
+    } else {
+      var fullPath = cwd.fullPath;
+      cwd.getDirectory(createFromDir)
+      .then((DirectoryEntry dirEntry) {
+        cwd = dirEntry;
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:198]: createDir() shouldn't also change the cwd because it limits the generality of the function.

File: gitnu/app/gitnufilesystem.dart [hunk 24]
```diff
@@ -0,0 +1,404 @@
...
+      if (dashP) {
+        var folders = dirName.split('/');
+        // Throw out './' or '/' if present on the beginning of our path.
+        if (folders[0] == '.' || folders[0] == '') {
+          folders.removeAt(0);
+        }
+
+        // If '/' is present then we change directories in createDir.
+        if (dirName[0] == "/") {
+          createDir(cwd, folders, display, dirName[0], "mkdir");
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:235]: Passing "/" as the sub-path here isn't the same as omitting it? Won't both resolve to root?

File: gitnu/app/gitnufilesystem.dart [hunk 25]
```diff
@@ -0,0 +1,404 @@
...
+        } else {
+          createDir(cwd, folders, display);
+        }
+      } else {
+        cwd.createDirectory(dirName, exclusive: true)
+        .then((_) {}, onError: (FileError error) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:241]: .then() on line above.

File: gitnu/app/gitnufilesystem.dart [hunk 26]
```diff
@@ -0,0 +1,404 @@
...
+          invalidOpForEntryType(error, "mkdir", dirName, display);
+        });
+      }
+    }
+  }
+
+  void updateFilename(String cmd, List<String> args, Function display, Function action) {
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:248]: comment for this method

File: gitnu/app/gitnufilesystem.dart [hunk 27]
```diff
@@ -0,0 +1,404 @@
...
+  void mvCommand(String cmd, List<String> args, Function display) {
+    void moveTo(srcDirEntry, destDirEntry, [name = ""]) {
+      if (name.isEmpty) {
+        srcDirEntry.moveTo(destDirEntry);
+      } else {
+        srcDirEntry.moveTo(destDirEntry, name);
+      }
+    };
+    
+    window.console.debug("here2");
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:314]: remove debugging

File: gitnu/app/gitnufilesystem.dart (hunk 20 above)
Comment (camfitz) [inline gitnu/app/gitnufilesystem.dart:171]: Hahaha - had this in to see if the loop case ever actually gets called (more than the obvious once to addAll). All gone.

File: gitnu/app/gitnufilesystem.dart [hunk 28]
```diff
@@ -0,0 +1,404 @@
...
+    // Moving to a folder? (e.g. second arg ends in '/').
+    if (dest[dest.length - 1].endsWith('/')) {
+      cwd.getDirectory(src)
+      .then((DirectoryEntry srcDirEntry) {
+            // Create blacklist for dirs we can't re-create.
+            var create = ['.', './', '..', '../', '/'].indexOf(dest) != -1 ? false : true;
+
+            if (create) {
+              cwd.createDirectory(dest)
+              .then((DirectoryEntry destDirEntry) => action(srcDirEntry, destDirEntry),
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:268]: .then()s in this method too

File: gitnu/app/gitnufilesystem.dart [hunk 29]
```diff
@@ -0,0 +1,404 @@
...
+
+    String src = args[0];
+    String dest = args[1];
+
+    // Moving to a folder? (e.g. second arg ends in '/').
+    if (dest[dest.length - 1].endsWith('/')) {
+      cwd.getDirectory(src)
+      .then((DirectoryEntry srcDirEntry) {
+            // Create blacklist for dirs we can't re-create.
+            var create = ['.', './', '..', '../', '/'].indexOf(dest) != -1 ? false : true;
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:264]: nit: sort of a double negative here. Would be better phrased as

var create = ['.', './', '..', '../', '/'].indexOf(dest) == -1;

File: gitnu/app/gitnufilesystem.dart (hunk 23 above)
Comment (camfitz) [inline gitnu/app/gitnufilesystem.dart:198]: Agreed. And that sits outside the mkdir behaviour anyways!

File: gitnu/app/gitnufilesystem.dart (hunk 24 above)
Comment (camfitz) [inline gitnu/app/gitnufilesystem.dart:235]: This was one of the existing functions from the example code.

I've noted the limitations of these (i.e. this won't actually properly handle mkdir /folderName - fires an exception, but am planning to fix in the future so I can make a start on Git.

Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:235]: Yep, SGTM - in terms of priorities Git stuff is way more important than getting things like cp/mv/mkdir working perfectly.

File: gitnu/app/gitnufilesystem.dart [hunk 30]
```diff
@@ -0,0 +1,404 @@
...
+  }
+
+  void printDirectory(Function display) {
+    display(StaticToolkit.htmlEscape(cwd.fullPath));
+  }
+
+  void rmCommand(String cmd, List<String> args, Function display) {
+    // Remove recursively? If so, remove the flag(s) from the arg list.
+    var recursive = false;
+    var switches = ['-r', '-f', '-rf', '-fr'];
```
Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:359]: Fancy way of doing this:

```
var originalLength = args.length;
args.removeWhere((arg) => switches.contains(arg));
var recursive = args.length != originalLength;
```

Comment (soulplant) [inline gitnu/app/gitnufilesystem.dart:359]: Also, the -f argument to rm is for force, ie: don't ask if it should delete something, so "rm -f" is not a recursive rm.

File: gitnu/app/gitnuterminal.dart [hunk 31]
```diff
@@ -130,16 +110,19 @@ class GitnuTerminal extends Terminal {
       String cmd = "";
       if (!cmdline.isEmpty) {
         cmdline.trim();
-        args = htmlEscape(cmdline).split(' ');
+        args = StaticToolkit.htmlEscape(cmdline).split(' ');
```
Comment (soulplant) [inline gitnu/app/gitnuterminal.dart:113]: Why do we need to htmlEscape() the user input?

File: gitnu/app/gitnuterminal.dart [hunk 32]
```diff
@@ -149,26 +132,62 @@ class GitnuTerminal extends Terminal {
...
+  /**
+   * Handles commands entered previously and redisplaying them in the input field
+   * when the up and down arrows are used.
+   */
+  void historyHandler(KeyboardEvent event) {
+    var histtemp = "";
+    int upArrowKey = 38;
+    int downArrowKey = 40;
+
+    /* keyCode == up-arrow || keyCode == down-arrow */
```
Comment (soulplant) [inline gitnu/app/gitnuterminal.dart:144]: This comment is unnecessary.

File: gitnu/app/gitnuterminal.dart [hunk 33]
```diff
@@ -149,26 +132,62 @@ class GitnuTerminal extends Terminal {
...
+
+      // Up or down
+      if (historyPosition < history.length) {
+        history[historyPosition] = input.value;
+      } else {
+        histtemp = input.value;
+      }
+    }
+
+    if (event.keyCode == upArrowKey) { // Up-arrow keyCode
```
Comment (soulplant) [inline gitnu/app/gitnuterminal.dart:156]: Unnecessary comment.

File: gitnu/app/gitnuterminal.dart [hunk 34]
```diff
@@ -149,26 +132,62 @@ class GitnuTerminal extends Terminal {
...
+      historyPosition--;
+      if (historyPosition < 0) {
+        historyPosition = 0;
+      }
+    } else if (event.keyCode == downArrowKey) { // Down-arrow keyCode
```
Comment (soulplant) [inline gitnu/app/gitnuterminal.dart:161]: Unnecessary comment.

File: gitnu/app/gitnuterminal.dart [hunk 35]
```diff
@@ -149,26 +132,62 @@ class GitnuTerminal extends Terminal {
...
+      historyPosition++;
+      if (historyPosition >= history.length) {
+        historyPosition = history.length - 1;
+      }
+    }
+
+    /* keyCode == up-arrow || keyCode == down-arrow */
```
Comment (soulplant) [inline gitnu/app/gitnuterminal.dart:168]: Unnecessary comment.

File: gitnu/app/gitnuterminal.dart [hunk 36]
```diff
@@ -149,26 +132,62 @@ class GitnuTerminal extends Terminal {
...
+    if (event.keyCode == upArrowKey || event.keyCode == downArrowKey) {
+      // Up or down
```
Comment (soulplant) [inline gitnu/app/gitnuterminal.dart:170]: Unnecessary comment.

File: gitnu/app/gitnuterminal.dart [hunk 37]
```diff
@@ -149,26 +132,62 @@ class GitnuTerminal extends Terminal {
     }
   }
   
-  void initializeFilesystem(bool persistent, int size) {
+  /**
+   * Handles commands entered previously and redisplaying them in the input field
+   * when the up and down arrows are used.
+   */
+  void historyHandler(KeyboardEvent event) {
+    var histtemp = "";
```
Comment (soulplant) [inline gitnu/app/gitnuterminal.dart:140]: I don't think this variable is necessary, you can just refer to input.value (it's only assigned once and read once).

File: gitnu/app/gitnufilesystem.dart (hunk 7 above)
Comment (camfitz) [inline gitnu/app/gitnufilesystem.dart:51]: Factored. Now runs an optional callback and root is set elsewhere.

File: gitnu/app/gitnufilesystem.dart (hunk 17 above)
Comment (camfitz) [inline gitnu/app/gitnufilesystem.dart:131]: Done! I like this.

File: gitnu/app/gitnufilesystem.dart (hunk 18 above)
Comment (camfitz) [inline gitnu/app/gitnufilesystem.dart:148]: Changed to _cwd, _root (private library variables). Added getters and setters, as well as string getters.

File: gitnu/app/gitnufilesystem.dart (hunk 21 above)
Comment (camfitz) [inline gitnu/app/gitnufilesystem.dart:180]: I think, for the most part, I can infer the cmd from the FileSystem alone.

File: gitnu/app/gitnufilesystem.dart (hunk 30 above)
Comment (camfitz) [inline gitnu/app/gitnufilesystem.dart:359]: removed -f, subbed in removeWhere, love it.

File: gitnu/app/gitnuterminal.dart (hunk 31 above)
Comment (camfitz) [inline gitnu/app/gitnuterminal.dart:113]: Good point.
